### DISCUSSION TITLE
Debug library not accessible: `attempt to index a nil value (global 'debug')`

### DISCUSSION THREAD

**Discussion by ashvinsharma** [mlua-rs/mlua#448] [Q&A]
I am trying to use [lua-http](https://github.com/daurnimator/lua-http) in [vector](https://github.com/vectordotdev/vector), but I got this error- 

```shell
ERROR vector::topology::builder: Configuration error. error=Transform "my_transform_id": Cannot evaluate Lua code in "source": runtime error: /usr/local/share/lua/5.4/http/bit.lua:14: attempt to index a nil value (global 'debug')
stack traceback:
    [C]: in metamethod 'index'
    /usr/local/share/lua/5.4/http/bit.lua:14: in main chunk
    [C]: in function 'require'
    /usr/local/share/lua/5.4/http/h2_connection.lua:7: in main chunk
    [C]: in function 'require'
    /usr/local/share/lua/5.4/http/client.lua:8: in main chunk
    [C]: in function 'require'
    /usr/local/share/lua/5.4/http/request.lua:5: in main chunk
    [C]: in function 'require'
    /etc/vector-lua-scripts/test.lua:2: in function 'test'
    /etc/vector-lua-scripts/test.lua:36: in main chunk
    [C]: in function 'require'
    [string "src/transforms/lua/v2/mod.rs:263:17"]:1: in main chunk

    [C]: in function 'require'

    (command line):1: in main chunk

    [C]: in ?
```

I debugged a little and found it broke at this [line](https://github.com/daurnimator/lua-http/blob/master/http/bit.lua#L14) 
```lua
-- file: http/bit.lua
local info = debug.getinfo(1, "Sl")
```
Can I do anything to make mlua have access to debug std library? 

#### NOTE
I tried to run this lua-http library in a clean docker image containing lua5.4 and it runs perfectly fine. Here's that dockerfile- 
```Dockerfile
FROM timberio/vector:0.34.1-debian

RUN apt-get update -y && \
    apt-get upgrade -y && \
    apt-get install -y lua5.4 liblua5.4-dev libssl-dev build-essential wget && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

RUN wget https://luarocks.org/releases/luarocks-3.8.0.tar.gz --no-check-certificate && \
    tar zxpf luarocks-3.8.0.tar.gz && cd luarocks-3.8.0 && \
    ./configure --lua-version=5.4 && \
    make && make install

RUN luarocks install http --lua-version=5.4 --tree=/usr/local
RUN ["vector", "--version"]

ENTRYPOINT ["/usr/bin/vector"]
```

And I ran the following commands 
```shell
root@a4cfc0f2e9eb:/# lua5.4
Lua 5.4.4  Copyright (C) 1994-2022 Lua.org, PUC-Rio
> request = require "http.request"
> req = request.new_from_uri("http://google.com")
> for k, v in req.headers:each() do
        print(k, v)
    end
:method	GET
:authority	google.com
:path	/
:scheme	http
user-agent	lua-http/0.4
> print(debug)
table: 0xaaaad1f8e770
```

This would otherwise break on the first line with mlua.

**Comment by khvzak** [comment]
> Can I do anything to make mlua have access to debug std library?

`debug` library is unsafe and improper use can cause segfaults on Rust side. This is the reason why it's not available by default. You need to create a Lua instance using [`Lua::unsafe_new`](https://docs.rs/mlua/latest/mlua/struct.Lua.html#method.unsafe_new) to load `debug` library.

**Reply by ashvinsharma** [reply]
But this means I have to modify vector’s source code for this change, right? I think i will switch my library instead 😅.

**Reply by khvzak** [reply]
Seems vector explicitly disabled debug library:
See:
```rust
        // In order to support loading C modules in Lua, we need to create unsafe instance
        // without debug library.
        let lua = unsafe {
            mlua::Lua::unsafe_new_with(mlua::StdLib::ALL_SAFE, mlua::LuaOptions::default())
        };
```
https://github.com/vectordotdev/vector/blob/9693f5253c2ce306de3cfa40d0c52439df98fc71/src/transforms/lua/v1/mod.rs#L115-L119

However you can try the following workaround (mocking debug library):
```lua
debug = {
  getinfo = function()
     return {
         currentline = 0,
         source = "?"
    }
  end
}
```
The lua-http library does not really require `debug` to be functional

**Reply by ashvinsharma** [reply]
Okay I will try this. I have seen debug getting called multiple times in that library though. I’m not well equipped with lua skills to handle if something breaks. 😅
